pkg/base/replicated.go

```diff
@@ -152,7 +157,100 @@ func renderReplicated(u *upstreamtypes.Upstream, renderOptions *RenderOptions) (
 		}
 	}
 
-	return &commonBase, helmBases, nil
+	for i, chart := range kotsHelmCharts {
+		chartBytes, err := yaml.Marshal(chart)
+		if err != nil {
+			return nil, nil, nil, errors.Wrapf(err, "failed to marshal helm chart %s", kotsHelmChartPaths[i])
+		}
+
+		renderedKotsKinds.HelmCharts = append(renderedKotsKinds.HelmCharts, &kotsutil.KotsKindsFile{Path: kotsHelmChartPaths[i], Content: chartBytes})
+	}
+
+	return &commonBase, helmBases, renderedKotsKinds, nil
+}
+
+func renderKotsKinds(upstreamFiles []upstreamtypes.UpstreamFile, renderedConfig *kotsv1beta1.Config, renderOptions *RenderOptions, builder *template.Builder) (*kotsutil.KotsKindsBundle, error) {
+	kotsKinds := kotsutil.KotsKindsBundle{}
+
+	for _, upstreamFile := range upstreamFiles {
+		document := &Document{}
+		if err := yaml.Unmarshal(upstreamFile.Content, document); err != nil {
+			continue
+		}
+
+		decode := scheme.Codecs.UniversalDeserializer().Decode
+		_, gvk, err := decode(upstreamFile.Content, nil, nil)
+		if err != nil {
+			if document.APIVersion == "kots.io/v1beta1" {
+				return nil, errors.Wrapf(err, "failed to decode kots yaml %s", upstreamFile.Path)
+			}
+			continue
+		}
+
+		switch gvk.String() {
+		case "kots.io/v1beta1, Kind=Config":
+			// Use the rendered Config instead of the upstream.
+			configBytes, err := yaml.Marshal(renderedConfig)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to marshal rendered config")
+			}
+			kotsKinds.Config = &kotsutil.KotsKindsFile{Path: upstreamFile.Path, Content: configBytes}
+			continue
+
+		case "kots.io/v1beta1, Kind=ConfigValues":
+			// Skip the ConfigValues because the Config section was already rendered.
+			continue
+
+		case "kots.io/v1beta1, Kind=HelmChart":
+			// Skip HelmCharts because they get rendered later.
```

Comment:
can't we move rendering the HelmChart custom resources here and have them be used later instead of rendering and adding them later?